### ISSUE TITLE
[Feature Request] Mean Square Loss Function 

### ISSUE THREAD

**Issue by kestrelm** [davisking/dlib#152]
Hello,

It will be great if Mean Square Loss was part of your default library (For Linear Regression).
I know you can write your own loss layers but this is a very commonly used feature for DNN so it will be fantastic if it could be included by default.

Can you give a brief description of how the gradients are calculated in your API for the loss layer class?
For Mean Square Loss, I am assuming the local gradient should be calculated something like this:
https://github.com/nyanp/tiny-cnn/blob/master/tiny_cnn/lossfunctions/loss_function.h (under mse)

How do we go about implementing Mean Square Loss ourselves in your API? I read your docs about it but it wasn't clear about the gradient calculation portion.

Thanks!

**Comment by davisking** [comment]
Indeed, that's something that will be added before the next release.  

The instructions for adding more loss layers are here: http://dlib.net/dlib/dnn/loss_abstract.h.html#EXAMPLE_LOSS_LAYER_

**Comment by dennisfrancis** [comment]
This my first time here. If nobody else is working on this, I would like to give this a shot. Thanks.

**Comment by davisking** [comment]
Go for it.  I've gotten side tracked into other things so I'm not adding it.  It's easy to do though.

**Comment by dennisfrancis** [comment]
Hi @davisking, I created a loss_mean_squared loss layer in my forked repo and tried running a very simple one variable linear regression ( y = mx + c ) using it (see https://gist.github.com/dennisfrancis/444dd3e918f61f898f9d1551eeb8cde0 ). Noticed that the slope coefficient is calculated correctly, but not the intercept term. I am using a fc layer before the loss layer which is supposed to add bias terms as I understand from the documentation. 

I also tried using fc_no_bias instead with manual insertion of 1's in the input data to act as bias terms, but did not work. Could you please give me some hint on what I may be missing ? Thanks !

**Comment by davisking** [comment]
I don't know.  Maybe you aren't seeing what you expect because of weight decay, because the gradients are wrong, or some other thing.  I'm too busy right now to look over your repository, test your code, and figure it out :)

**Comment by dennisfrancis** [comment]
Ok, I figured that out. What I was missing was the learning rate multiplier for bias needs to be tuned separately. I will soon give a pull request which implements the mse loss layer with test cases.
